File: ios/Cloudmoji/Cloudmoji/Views/CloudHomeButton.swift
import SwiftUI

enum HomeButtonMetrics {
    /// A **child** taps this — it is how he gets out of a mini-app he did not
    /// mean to open — so it takes the 64pt floor, not the 44pt parent-chrome
    /// one. It is the only navigation control left in the app.
    ///
    /// 84, not 64, and centred rather than tucked in a corner. Tested on a real
    /// phone with the person it is for: 64pt in the bottom-left was hard to
    /// *find*, which is a different failure from hard to hit. The corner is
    /// where adults expect Back; a two-year-old scans the middle. This is the
    /// most important control in the app for a child who has opened the wrong
    /// thing, so it gets the most generous target in the app.
    static let side: CGFloat = 84

    /// The mascot inside the circle. Sized so the cloud's letterboxed art
    /// (`size × size * 0.78`) sits inside the plate with a ring around it.
    static let glyphSize: CGFloat = 58

    /// How far the hosting screen keeps its own content clear of the button:
    /// the button, its inset, and a little air.
    static let reservedHeight: CGFloat = 108

    static let inset: CGFloat = 12

    /// Design system Active States: tabs `scale(0.9)`.
    static let pressedScale: CGFloat = 0.9

    static let borderWidth: CGFloat = 2
}

/// The way out of every mini-app.
///
/// A cloud rather than a chevron, and 64pt rather than a nav bar's 44, because
/// the person who most needs it cannot read "Back" and does not know that the
/// top-left corner of a screen means anything. He does know the cloud.
///
/// Its own type so the 64pt rule can be measured off a real render rather than
/// read back out of the constant that set it.
struct CloudHomeButton: View {
    /// The newer mini-apps tint the navigation ring with their own identity.
    /// `nil` preserves the original Words / Count treatment exactly.
    var accent: Color?
    let action: () -> Void

    init(accent: Color? = nil, action: @escaping () -> Void) {
        self.accent = accent
        self.action = action
    }

    var body: some View {
        Button {
            Haptics.tap()
            action()
        } label: {
            ZStack(alignment: .bottomTrailing) {
                CloudMascot(mood: .happy, size: HomeButtonMetrics.glyphSize)
                    // Hidden, unlike every other mascot in the app.
                    //
                    // `CloudMascot` publishes `mascot-<mood>` as its identifier, and
                    // that is the only way `WordsModeUITests` can observe a
                    // celebration. A second, permanently-happy cloud on the same
                    // screen would make every one of those lookups ambiguous — the
                    // test would measure whichever matched first and could pass with
                    // the celebration deleted. The button's own label is what
                    // VoiceOver reads here.
                    .accessibilityHidden(true)

                Image(systemName: "house.fill")
                    .font(.system(size: 12, weight: .black))
                    .foregroundStyle(Theme.bgPrimary)
                    .frame(width: 26, height: 26)
                    .background(accent ?? Theme.teal, in: Circle())
                    .overlay(Circle().stroke(Color.white.opacity(0.55), lineWidth: 1))
                    .offset(x: 3, y: 3)
                    .accessibilityHidden(true)
            }
                .frame(width: HomeButtonMetrics.side, height: HomeButtonMetrics.side)
                // Two layers, and the opaque one is load-bearing. The button
                // floats over a screen that scrolls underneath it — the animal
                // grid and the emoji grid both do — and `Theme.surface` is white
                // at 4%, so on its own the cloud came out sitting on top of a
                // half-visible emoji. The near-opaque plate is the one the tab
                // bar used to wear, at the same 0.95, so the two read as the
                // same piece of chrome in the same corner.
                .background(Theme.bgPrimary.opacity(0.95), in: Circle())
                .background((accent ?? Theme.textPrimary).opacity(accent == nil ? 0.04 : 0.14), in: Circle())
                .overlay(
                    Circle().stroke(
                        accent?.opacity(0.44) ?? Theme.surfaceBorderStrong,
                        lineWidth: HomeButtonMetrics.borderWidth
                    )
                )
                .shadow(color: accent?.opacity(0.18) ?? .clear, radius: 14, y: 5)
                // Without this the hit area is the mascot's own art box and the
                // ring of plate a toddler aims at is dead. `.contentShape` takes
                // the circle rather than a rectangle so the corners outside the
                // plate do not swallow taps meant for what is underneath.
                .contentShape(Circle())
        }
        .buttonStyle(PressScale(scale: HomeButtonMetrics.pressedScale))
        .accessibilityLabel("Home")
        .accessibilityIdentifier("home-btn")
    }
}

#Preview("Home button") {
    ZStack {
        Theme.background.ignoresSafeArea()
        HStack {
            CloudHomeButton {}
            CloudHomeButton(accent: Theme.coral) {}
        }
    }
}

File: android/app/src/main/java/app/cloudmoji/android/ui/common/CloudHomeButton.kt
package app.cloudmoji.android.ui.common

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.role
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import app.cloudmoji.android.model.MascotMood
import app.cloudmoji.android.ui.mascot.CloudMascot
import app.cloudmoji.android.ui.theme.BackgroundPrimary
import app.cloudmoji.android.ui.theme.SurfaceBorder
import app.cloudmoji.android.ui.theme.Teal
import app.cloudmoji.android.ui.theme.TextPrimary

/** Every number [CloudHomeButton] is drawn from. Mirrors iOS `CloudHomeButton.swift`'s `HomeButtonMetrics`. */
object HomeButtonMetrics {
    /** A **child** taps this — it is how they get out of a mini-app they did
     * not mean to open — so it takes 84dp, well past the 64dp floor for
     * anything else a child taps, and sits centred rather than tucked in a
     * corner: a corner is where an adult expects Back, and a toddler scans
     * the middle. This is the most important control in the app for a child
     * who has opened the wrong thing, so it gets the most generous target. */
    val side = 84.dp
    val mascotSize = 58.dp

    /** How far the hosting screen keeps its own content clear of the button —
     * the button, its inset, and a little air. [MiniAppScaffold] reserves
     * this band rather than relying on safe-area padding, which only insets
     * where a scroll view's content *ends*: everything in the middle would
     * otherwise travel underneath the cloud. */
    val reservedHeight = 108.dp
    val inset = 12.dp

    /** Design system Active States: tabs `scale(0.9)`. */
    const val pressedScale = 0.9f

    val borderWidth = 2.dp
    val badgeSize = 26.dp
}

/**
 * The way out of every mini-app: a cloud, not a chevron, because the person
 * who most needs it cannot read "Back" and does not know the top-left corner
 * of a screen means anything — but does know the cloud. Ported from iOS
 * `CloudHomeButton.swift`.
 *
 * [accent] tints the little house badge; `null` uses the app's own teal
 * rather than a per-mini-app identity color, since Words has none of its own.
 */
@Composable
fun CloudHomeButton(
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    accent: Color? = null,
) {
    val interactionSource = remember { MutableInteractionSource() }
    val badgeTint = accent ?: Teal

    Box(
        contentAlignment = Alignment.Center,
        modifier = modifier
            .size(HomeButtonMetrics.side)
            .pressScale(interactionSource, HomeButtonMetrics.pressedScale)
            // Two layers, and the near-opaque one is load-bearing: this
            // button floats over content that scrolls underneath it (the
            // emoji grid), and a translucent-only background would leave the
            // cloud sitting on top of a half-visible tile.
            .shadow(14.dp, CircleShape)
            .background(BackgroundPrimary.copy(alpha = 0.95f), CircleShape)
            .background((accent ?: TextPrimary).copy(alpha = if (accent == null) 0.04f else 0.14f), CircleShape)
            .border(HomeButtonMetrics.borderWidth, accent?.copy(alpha = 0.44f) ?: SurfaceBorder, CircleShape)
            .clip(CircleShape)
            .clickable(
                interactionSource = interactionSource,
                indication = null,
                role = Role.Button,
                onClick = onClick,
            )
            .semantics {
                role = Role.Button
                contentDescription = "Home"
            }
            .testTag("cloud-home"),
    ) {
        // Hidden from the accessibility tree, unlike every other mascot in
        // the app: `CloudMascot` publishes its own "Cloudmoji" node, and a
        // second one nested inside this button's "Home" node would surface
        // as an ambiguous, separately-focusable stop for TalkBack — the same
        // reason iOS marks this instance `.accessibilityHidden(true)`.
        Box(modifier = Modifier.clearAndSetSemantics {}) {
            CloudMascot(mood = MascotMood.Happy, size = HomeButtonMetrics.mascotSize)
        }

        Box(
            contentAlignment = Alignment.Center,
            modifier = Modifier
                .align(Alignment.BottomEnd)
                .offset(x = 3.dp, y = 3.dp)
                .size(HomeButtonMetrics.badgeSize)
                .background(badgeTint, CircleShape)
                .border(1.dp, Color.White.copy(alpha = 0.55f), CircleShape),
        ) {
            Text(text = "🏠", fontSize = 13.sp)
        }
    }
}
